src/packet/buffer_tx.rs

```diff
@@ -396,7 +359,342 @@ impl fmt::Debug for Packetized {
+            .1;
+        Some((seq_no, self.get_cached_packet_by_seq_no(seq_no)?))
+    }
+
+    fn remove_old_packets(&mut self, now: Instant) {
+        if self.evict_in_batches {
+            if let Some(first_seq_no_thats_not_too_old) =
+                self.find_first_seq_no_thats_not_too_old(now)
+            {
+                if let Some(first_seq_no_thats_not_too_old) = first_seq_no_thats_not_too_old {
+                    self.packet_by_seq_no = self
+                        .packet_by_seq_no
+                        .split_off(&first_seq_no_thats_not_too_old);
+                } else {
+                    // They are all too old
+                    self.packet_by_seq_no.clear();
+                }
+            }
+        } else {
+            while let Some(first_seq_no_thats_too_old) = self.find_first_seq_no_thats_too_old(now) {
+                self.packet_by_seq_no.remove(&first_seq_no_thats_too_old);
+            }
+        }
+    }
+
+    fn find_first_seq_no_thats_too_old(&self, now: Instant) -> Option<SeqNo> {
+        if self.packet_by_seq_no.len() > self.max_packet_count {
+            let first_seq_no = self.packet_by_seq_no.keys().next()?;
+            // Too old because of max_packet_count.
+            return Some(*first_seq_no);
+        }
+        // If the max_packet_age is so old that checked_sub returns None, we shouldn't remove based on max_packet_age.
+        let min_queued_at = now.checked_sub(self.max_packet_age)?;
+
+        let (first_seq_no, first_packet) = self.packet_by_seq_no.iter().next()?;
+        if first_packet.queued_at <= min_queued_at {
+            // Too old because of max_packet_age
+            return Some(*first_seq_no);
+        }
+        None
+    }
+
+    // None == nothing is too old
+    // Some(None) == everything is too old
+    // Some(Some(seq_no)) == everything before this is too old
+    fn find_first_seq_no_thats_not_too_old(&self, now: Instant) -> Option<Option<SeqNo>> {
+        let too_many_packets_count = self
+            .packet_by_seq_no
+            .len()
+            .saturating_sub(self.max_packet_count);
+        if too_many_packets_count > 0 {
+            return Some(
+                self.packet_by_seq_no
+                    .keys()
+                    .nth(too_many_packets_count)
+                    .cloned(),
+            );
+        }
+
+        // If the max_packet_age is so old that checked_sub returns None, we shouldn't remove based on max_packet_age.
+        let min_queued_at = now.checked_sub(self.max_packet_age)?;
+
+        // There is no packet, so I guess we'll clear it.  But that's a no-op anyway.
+        let Some(first_packet) = self.packet_by_seq_no.values().next() else {
+            return Some(None)
+        };
+
+        if first_packet.queued_at <= min_queued_at {
+            return Some(
+                self.packet_by_seq_no
+                    .iter()
+                    .find(|(_, packet)| packet.queued_at > min_queued_at)
```

Comment:
I believe it’s safer to evict based on media time (rtp time). For example in a SFU scenario It protects us from jitter / latency / drift making the buffer independent from those. A couple of extremes just for reasoning
- 1s worth of media can be queued in 20ms and vice versa
- In a SFU scenario (origin -> SFU -> palyer) if the origin drifts behind, the player slows down, the SFU should not shorten the buffer time, and vice versa
- with screenshares you have large frames (high latency), and very low fps (down to one every few secs). There is a higher chance for the player to nack an evicted buffer which would result in PLIs

Reply:
For a relatively large time duration (say, 3 seconds), it doesn't really matter if you're using the RTP timestamp or the send time.  After that long of a time, it won't be useful either way. 

If you use the RTP time, then you have to do clock synchronization between the SFU's clock and the RTP timestamps, and that can be difficult.  And even if you get it right, the net result is that you'll be choosing to keep packets in the RTX cache for less time in the case that they arrive late relative to the others.  So the RTX cache gets a little smaller?  If that's the goal, there are many other ways to do it that are better (ACK-based cache eviction, dynamic duration based on a histogram of NACK times, etc).  

So I don't really think it's worth the trouble of the clock synchronization when it's easier to just make the time duration somewhat large and have other mechanisms for keeping the RTX cache small.